File: 1091.shortest-path-in-binary-matrix/solution.go
const (
    empty = 0
    visited = 2
)

func shortestPathBinaryMatrix(grid [][]int) int {
    last := len(grid) - 1
    length := 0
    queue := [][2]int{{0, 0}}
    for len(queue) > 0 {
        length++
        for i := len(queue); i > 0; i-- {
            r, c := queue[0][0], queue[0][1]
            queue = queue[1:]
            if grid[r][c] != empty {
                continue
            }
            if r == last && c == last {
                return length
            }
            grid[r][c] = visited
            for rr := max(r - 1, 0); rr <= min(r + 1, last); rr++ {
                for cc := max(c - 1, 0); cc <= min(c + 1, last); cc++ {
                    if grid[rr][cc] == empty {
                        queue = append(queue, [2]int{rr, cc})
                    }
                }
            }
        }
    }
    return -1
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: 1091.shortest-path-in-binary-matrix/solution.kt
class Solution {
    fun shortestPathBinaryMatrix(grid: Array<IntArray>): Int {
        val N = grid.lastIndex
        var length = 0
        val queue = ArrayDeque<Pair<Int, Int>>()
        queue.offer(0 to 0)
        while (queue.isNotEmpty()) {
            length++
            repeat(queue.size) {
                val (r, c) = queue.poll()
                if (grid[r][c] == 0) {
                    grid[r][c] = 2
                    if (r == N && c == N) {
                        return length
                    }
                    for (rr in maxOf(0, r - 1) .. minOf(N, r + 1)) {
                        for (cc in maxOf(0, c - 1) .. minOf(N, c + 1)) {
                            if (grid[rr][cc] == 0) {
                                queue.offer(rr to cc)
                            } 
                        }
                    }
                }
            }
        }
        return -1
    }
}